File: processor/src/main/java/com/qualityminds/lazyval/processor/internal/codegen/beanvalidation/TemporalValidatorBuilder.java
package com.qualityminds.lazyval.processor.internal.codegen.beanvalidation;

import com.palantir.javapoet.*;
import com.qualityminds.lazyval.processor.internal.codegen.GeneratedStamp;
import com.qualityminds.lazyval.processor.spi.GeneratorResult;
import com.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import java.util.Set;
import java.util.stream.Stream;

class TemporalValidatorBuilder {

    private static final Set<String> TEMPORAL_TYPES = Set.of(
            "LocalDate", "LocalDateTime", "LocalTime",
            "OffsetDateTime", "OffsetTime", "ZonedDateTime",
            "Instant", "Year", "YearMonth", "HijrahDate",
            "JapaneseDate", "MinguoDate", "ThaiBuddhistDate"
    );

    private static final ClassName CONSTRAINT_VALIDATOR = ClassName.get("jakarta.validation", "ConstraintValidator");
    private static final ClassName CONSTRAINT_VALIDATOR_CONTEXT = ClassName.get("jakarta.validation", "ConstraintValidatorContext");

    static boolean supports(String typeName) {
        return TEMPORAL_TYPES.contains(typeName);
    }

    static Stream<GeneratorResult> build(ValidatedGeneratorElement element, String packageName) {
        return Stream.of(
                buildValidator(element, packageName, "Past"),
                buildValidator(element, packageName, "Future"),
                buildValidator(element, packageName, "PastOrPresent"),
                buildValidator(element, packageName, "FutureOrPresent")
        ).flatMap(s -> s);
    }

    private static Stream<GeneratorResult> buildValidator(ValidatedGeneratorElement element, String packageName, String annotationName) {
        TypeMirror lazyvalType = element.element().asType();
        String className = element.typeName().name() + annotationName + "Validator";

        ClassName temporalAnnotation = ClassName.get("jakarta.validation.constraints", annotationName);

        TypeSpec validator = TypeSpec.classBuilder(className)
                .addAnnotation(GeneratedStamp.forGenerator(BeanValidationGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(ParameterizedTypeName.get(
                        CONSTRAINT_VALIDATOR, temporalAnnotation, TypeName.get(lazyvalType)))
                .addMethod(buildIsValid(element, lazyvalType, annotationName))
                .build();

        JavaFile javaFile = JavaFile.builder(packageName, validator).build();
        return ResultHelper.toResultStream(javaFile, packageName, className);
    }

    private static MethodSpec buildIsValid(ValidatedGeneratorElement element, TypeMirror lazyvalType, String annotationName) {
        var builder = MethodSpec.methodBuilder("isValid")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(boolean.class)
                .addParameter(TypeName.get(lazyvalType), "value")
                .addParameter(CONSTRAINT_VALIDATOR_CONTEXT, "context")
                .beginControlFlow("if (value == null)")
                .addStatement("return true")
                .endControlFlow();

        String wrappedTypeName = element.wrappedType().typeName().simpleName();
        String nowExpression = resolveNowExpression(wrappedTypeName);

        if (annotationName.equals("Past") || annotationName.equals("Future")) {
            String comparison = annotationName.equals("Past") ? "isBefore" : "isAfter";
            builder.addStatement("return value.$L.$L($L)", element.accessor(), comparison, nowExpression);
        } else {
            String negatedMethod = annotationName.equals("PastOrPresent") ? "isAfter" : "isBefore";
            builder.addStatement("return !value.$L.$L($L)", element.accessor(), negatedMethod, nowExpression);
        }

        return builder.build();
    }

    private static String resolveNowExpression(String wrappedTypeName) {
        return switch (wrappedTypeName) {
            case "Instant" -> "java.time.Instant.now()";
            case "LocalDate" -> "java.time.LocalDate.now()";
            case "LocalDateTime" -> "java.time.LocalDateTime.now()";
            case "LocalTime" -> "java.time.LocalTime.now()";
            case "OffsetDateTime" -> "java.time.OffsetDateTime.now()";
            case "OffsetTime" -> "java.time.OffsetTime.now()";
            case "ZonedDateTime" -> "java.time.ZonedDateTime.now()";
            case "Year" -> "java.time.Year.now()";
            case "YearMonth" -> "java.time.YearMonth.now()";
            default -> "java.time.chrono." + wrappedTypeName + ".now()";
        };
    }
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/internal/codegen/beanvalidation/TemporalValidatorBuilder.kt
package com.qualityminds.lazyval.ksp.internal.codegen.beanvalidation

import com.qualityminds.lazyval.ksp.internal.codegen.GeneratedStamp.addGeneratedAnnotation
import com.qualityminds.lazyval.ksp.spi.Generator
import com.qualityminds.lazyval.ksp.spi.GeneratorResult
import com.qualityminds.lazyval.ksp.spi.ValidatedKspGeneratorElement
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import java.util.stream.Stream

internal object TemporalValidatorBuilder {

    private val TEMPORAL_TYPES = setOf(
        "LocalDate", "LocalDateTime", "LocalTime",
        "OffsetDateTime", "OffsetTime", "ZonedDateTime",
        "Instant", "Year", "YearMonth", "HijrahDate",
        "JapaneseDate", "MinguoDate", "ThaiBuddhistDate"
    )

    private val CONSTRAINT_VALIDATOR = ClassName("jakarta.validation", "ConstraintValidator")
    private val CONSTRAINT_VALIDATOR_CONTEXT = ClassName("jakarta.validation", "ConstraintValidatorContext")

    fun supports(typeName: String): Boolean = typeName in TEMPORAL_TYPES

    fun build(ctx: Generator.Context, element: ValidatedKspGeneratorElement, packageName: String): Stream<GeneratorResult> {
        return Stream.of(
            buildValidator(ctx, element, packageName, "Past"),
            buildValidator(ctx, element, packageName, "Future"),
            buildValidator(ctx, element, packageName, "PastOrPresent"),
            buildValidator(ctx, element, packageName, "FutureOrPresent")
        ).flatMap { it }
    }

    private fun buildValidator(ctx: Generator.Context, element: ValidatedKspGeneratorElement, packageName: String, annotationName: String): Stream<GeneratorResult> {
        val lazyvalTypeName = element.element.toClassName()
        val className = "${element.typeName.name}${annotationName}Validator"
        val temporalAnnotation = ClassName("jakarta.validation.constraints", annotationName)

        val typeSpec = TypeSpec.classBuilder(className)
            .addGeneratedAnnotation(BeanValidationGenerator::class, ctx)
            .addSuperinterface(
                CONSTRAINT_VALIDATOR.parameterizedBy(temporalAnnotation, lazyvalTypeName)
            )
            .addFunction(buildIsValid(element, lazyvalTypeName, annotationName))
            .build()

        val fileSpec = FileSpec.builder(packageName, className).addType(typeSpec).build()
        return ResultHelper.toResultStream(fileSpec, packageName, className)
    }

    private fun buildIsValid(element: ValidatedKspGeneratorElement, lazyvalTypeName: ClassName, annotationName: String): FunSpec {
        val wrappedTypeName = element.wrappedProperty.type.declaration.simpleName.asString()
        val nowExpression = resolveNowExpression(wrappedTypeName)

        val comparison = when (annotationName) {
            "Past" -> "return value.${element.kotlinAccessor}.isBefore($nowExpression)"
            "Future" -> "return value.${element.kotlinAccessor}.isAfter($nowExpression)"
            "PastOrPresent" -> "return !value.${element.kotlinAccessor}.isAfter($nowExpression)"
            "FutureOrPresent" -> "return !value.${element.kotlinAccessor}.isBefore($nowExpression)"
            else -> throw IllegalArgumentException("Unknown temporal annotation: $annotationName")
        }

        return FunSpec.builder("isValid")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addParameter("value", lazyvalTypeName.copy(nullable = true))
            .addParameter("context", CONSTRAINT_VALIDATOR_CONTEXT)
            .beginControlFlow("if (value == null)")
            .addStatement("return true")
            .endControlFlow()
            .addStatement(comparison)
            .build()
    }

    private fun resolveNowExpression(wrappedTypeName: String): String {
        return when (wrappedTypeName) {
            "Instant" -> "java.time.Instant.now()"
            "LocalDate" -> "java.time.LocalDate.now()"
            "LocalDateTime" -> "java.time.LocalDateTime.now()"
            "LocalTime" -> "java.time.LocalTime.now()"
            "OffsetDateTime" -> "java.time.OffsetDateTime.now()"
            "OffsetTime" -> "java.time.OffsetTime.now()"
            "ZonedDateTime" -> "java.time.ZonedDateTime.now()"
            "Year" -> "java.time.Year.now()"
            "YearMonth" -> "java.time.YearMonth.now()"
            else -> "java.time.chrono.$wrappedTypeName.now()"
        }
    }
}
